stan/math/rev/fun/generalized_inverse.hpp

```diff
@@ -0,0 +1,198 @@
+#ifndef STAN_MATH_REV_FUN_GENERALIZED_INVERSE_HPP
+#define STAN_MATH_REV_FUN_GENERALIZED_INVERSE_HPP
+
+#include <stan/math/rev/core.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/fun/cholesky_decompose.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/rev/fun/tcrossprod.hpp>
+#include <stan/math/prim/fun/crossprod.hpp>
+#include <stan/math/prim/fun/inverse_spd.hpp>
+#include <stan/math/rev/fun/mdivide_left_spd.hpp>
+#include <stan/math/prim/fun/mdivide_right_spd.hpp>
+#include <stan/math/rev/fun/inverse.hpp>
+
+namespace stan {
+namespace math {
+
+/*
+ * Reverse mode specialization of calculating the generalized inverse of a
+ * matrix.
+ *
+ * @param G specified matrix
+ * @return Generalized inverse of the matrix (an empty matrix if the specified
+ * matrix has size zero).
+ *
+ * @note Reverse mode differentiation algorithm reference:
+ *
+ * <ul><li> Golub, G.H. and Pereyra, V. The Differentiation of Pseudo-Inverses
+ * and Nonlinear Least Squares Problems Whose Variables Separate. <i>SIAM
+ * Journal on Numerical Analysis</i>, Vol. 10, No. 2 (Apr., 1973), pp.
+ * 413-432</li></ul>
+ *
+ * Equation 4.12 in the paper
+ *
+ *  See also
+ *  http://mathoverflow.net/questions/25778/analytical-formula-for-numerical-derivative-of-the-matrix-pseudo-inverse
+ *
+ *  Implementation is based on
+ *   Title: Uncertainties Python Package
+ *    Author: Eric O. LEBIGOT
+ *    Date: 2020
+ *    Availability:
+ * https://github.com/lebigot/uncertainties/blob/master/uncertainties/unumpy/core.py
+ */
+template <typename VarMat, require_rev_matrix_t<VarMat>* = nullptr>
+inline auto generalized_inverse(const VarMat& G) {
+  using value_t = value_type_t<VarMat>;
+  using ret_type = promote_var_matrix_t<VarMat, VarMat>;
+
+  if (G.size() == 0)
+    return ret_type(G);
+
+  if (G.rows() == G.cols())
+    return ret_type(inverse(G));
+
+  if (G.rows() < G.cols()) {
+    arena_t<VarMat> G_arena(G);
+    auto A_spd = tcrossprod(G_arena.val_op());
+    arena_t<VarMat> inv_G(
+        mdivide_left_spd(A_spd.val_op(), G_arena.val_op()).transpose());
+
+    arena_t<VarMat> res = inv_G;
```

Comment:
I was following the [inverse.hpp](https://github.com/stan-dev/math/blob/develop/stan/math/rev/fun/inverse.hpp) and there's something similar. On lines 31-33
```
  arena_t<T> arena_m = m;
  arena_t<promote_scalar_t<double, T>> res_val = arena_m.val().inverse();
  arena_t<ret_type> res = res_val;
```  
I see that the last line has `ret_type`. Do you know why `res_val` is repeated for `res`?

Reply:
That can be read as

```cpp
  arena_t<Eigen::Matrix<var, -1, -1>> arena_m = m;
  arena_t<Eigen::Matrix<double,  -1, -1>> res_val = arena_m.val().inverse();
  arena_t<Eigen::Matrix<var, -1, -1>> res = res_val;
```

So it's making the _values_ that are assigned to the output in `res_val`, but the actual vars for the output are not created tilll `res`. For yours you can read it as

```cpp
    arena_t<Eigen::Matrix<var, -1, -1>> inv_G(
        mdivide_left_spd(A_spd.val_op(), G_arena.val_op()).transpose());
    arena_t<Eigen::Matrix<var, -1, -1>> res = inv_G;
```

Where in `inv_G` you are making the full return type already. so there's no need to make res.

I think `inverse()` is making the values for the vars like that because it's trying to to an optimization


```cpp
 arena_t<promote_scalar_t<double, T>> res_val = arena_m.val().inverse();
  arena_t<ret_type> res = res_val;

  reverse_pass_callback([res, res_val, arena_m]() mutable {
    // notice res_val is called twice
    arena_m.adj() -= res_val.transpose() * res.adj_op() * res_val.transpose();
  });
```

It's because we use the values twice in `res_val`. but tbh idt this makes sense as an optimization because we're paying a big memory allocation but it's not like we get better instruction sets etc. by doing that. So I'd keep doing it the way your doing it, but just use `inv_G` and get rid of `res`

Reply:
thanks for the explanation, it's very helpful!